File: nodejs/data-api.js
'use strict';

var http = require('http'),
    q = require('q'),
    _ = require('underscore');

var DataApiClient = function(host, port, path) {

  var urls = {
    auth: path +'/ws/security/token',
    readContentId: path+'/ws/content/contentid/',
    readExternalId: path+'/ws/content/externalid/',
    create: path+'/ws/content',
    search: path+'/ws/search'
  }; 

  var makeRequest = function(method, path, token, payload, customHeaders) {
    var deferred = q.defer();
    var responseData = '';

    var params = {
      'hostname': host,
      'port': port,
      'method': method,
      'path': path,
      'headers': _.extend({ 'Content-Type': 'application/json' }, customHeaders)
    };

    if(token) {
      params.headers['X-Auth-Token'] = token;
    }

    var req = http.request(params, function(res) {
      var statusCode = res.statusCode;

      res.on('data', function(chunk) {
        responseData += chunk;
      });

      res.on('end', function() {
        if(statusCode === 401) {
          deferred.reject('401 Unauthorized');
	} else if(statusCode === 403) {
          deferred.reject('403 Forbidden');
	} else if(statusCode === 400) {
          deferred.reject('400 Bad Request');
	} else if(statusCode === 500) {
          deferred.reject('500 Internal Server Error');
	} else if(statusCode === 303) {
	  makeRequest(method, res.headers.location, token, payload, customHeaders).then(function(response) {
	    deferred.resolve(response);
	  }, function(error ) {
	    deferred.reject(error);
	  });
	} else {
          if(responseData && responseData !== '') {
            deferred.resolve(JSON.parse(responseData));
	  } else {
            deferred.resolve();
          }
	}
      });
    }); 

    req.on('error', function(error) {
      console.log('ERROR: '+ error);
      deferred.reject(error);
    });

    if(params.method === 'POST' || params.method === 'PUT') {
      req.write(JSON.stringify(payload));
    }

    req.end();

    return deferred.promise;
  };
  
  return {
    authenticate: function(username, password) {
      var payload = {
        'username': username,
	'password': password
      };

      return makeRequest('POST', urls.auth, undefined, payload);
    },

    invalidateToken: function(token) {
      return makeRequest('DELETE', urls.auth, token);
    },
    
    search: function(token, index, searchExpression, variant, rows) {
      var path = urls.search+'/public/select?q='+ searchExpression +'&wt=json';
      
      if(rows) path += '&rows='+rows;
      if(variant) path += '&variant='+variant;

      return makeRequest('GET', path, token);
    },

    create: function(token, variant, payload) {
      return makeRequest('POST', urls.create+'?variant='+variant, token, payload);
    },

    read: function(token, contentId, variant) {
      var path = (/\d+\.\d+/.test(contentId) ? urls.readContentId : urls.readExternalId) + contentId;

      if(variant) path += "?variant="+ variant;

      return makeRequest('GET', path, token);
    },

    update: function(token, contentId, variant, payload) {
      var path = (/\d+\.\d+/.test(contentId) ? urls.readContentId : urls.readExternalId) + contentId;

      if(variant) {
        path += "?variant="+ variant;
      }

      return makeRequest('PUT', path, token, payload, { 'If-Match': payload.id +'.'+ payload.version });
    }
  };
};

module.exports = {
  client: DataApiClient
};

File: python/dataapi.py
#!/usr/bin/python

import urllib
import httplib
import json
import re

class Client:

    _auth = "/security/token"
    _readContentId = "/content/contentid/"
    _readExternalId = "/content/externalid/"
    _create = "/content"
    _search = "/search"

    def __init__(self, host, port, path):
        self._host = host
        self._port = port
        self._path = path

    def _makeRequest(self, method, path, token=None, payload=None, customHeaders=None):
        connection = httplib.HTTPConnection(self._host, self._port)

        headers = {
          "Content-Type": "application/json",
        }

        if customHeaders != None: headers.update(customHeaders)
        if token != None: headers['X-Auth-Token'] = token

        connection.request(
            method=method,
            url=path,
            body=json.dumps(payload),
            headers=headers
        )

        response = connection.getresponse()

        if response.status == 204:
          return None, response.getheader('ETag')
        else:
          jsonData = json.loads(response.read())
          if response.status in [401, 403, 404, 400, 409, 500]:
              raise Exception("HTTP {0}: {1}".format(response.status, jsonData["message"]))
          elif response.status == 303:
              return self._makeRequest(method,
                                       jsonData["location"],
                                       token,
                                       payload,
                                       customHeaders)
          else:
              return jsonData, response.getheader('ETag')

    def authenticate(self, username, password):
        payload = {
            "username": username,
            "password": password
        }

        return self._makeRequest(
            "POST",
            self._path+self._auth,
            payload=payload
        )[0]

    def invalidateToken(self, token):
        return self._makeRequest(
            "DELETE",
            self._path+self._auth,
            token
        )[0]

    def search(self, token, index, searchExpression, variant=None, rows=None):
        path = "{0}{1}/{2}/select?q={3}&wt=json".format(
            self._path,
            self._search,
            urllib.quote(index),
            urllib.quote(searchExpression)
        )

        if rows != None: path += "&rows={0}".format(urllib.quote(rows))
        if variant != None: path += "&variant={0}".format(urllib.quote(variant))

        return self._makeRequest(
            "GET",
            path,
            token
        )[0]

    def create(self, token, payload, variant):
        path = self._path+self._create

        if variant:
            path += "?variant={0}".format(variant)

        return self._makeRequest(
            "POST",
            path,
            token,
            payload
        )

    def read(self, token, contentId, variant):
        path = ""
        if re.search("\\d+\\.\\d+", contentId) != None:
            path = self._path+self._readContentId
        else:
            path = self._path+self._readExternalId

        path += contentId
        if variant:
            path += "?variant={0}".format(variant)

        return self._makeRequest(
            "GET",
            path,
            token
        )

    def update(self, token, payload, etag, variant):
        path = ""
        if re.search("\\d+\\.\\d+", payload["id"]) != None:
            path = self._path+self._readContentId
        else:
            path = self._path+self._readExternalId

        path += payload["id"]
        if variant:
            path += "?variant={0}".format(variant)

        return self._makeRequest(
            "PUT",
            path,
            token,
            payload,
            { "If-Match": "{0}".format(etag) }
        )
